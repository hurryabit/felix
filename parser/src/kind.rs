#[allow(non_camel_case_types)]
#[allow(clippy::upper_case_acronyms)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, logos::Logos)]
#[repr(u16)]
pub enum SyntaxKind {
    // Terminals aka tokens
    #[token("fun")]
    FUN = 0,
    #[token("let")]
    LET,
    #[token("rec")]
    REC,
    #[token("in")]
    IN,
    #[token("if")]
    IF,
    #[token("then")]
    THEN,
    #[token("else")]
    ELSE,
    #[token("true")]
    TRUE,
    #[token("false")]
    FALSE,

    #[token("->")]
    ARROW,
    #[token("=")]
    ASSIGN,
    #[token("(")]
    LPAREN,
    #[token(")")]
    RPAREN,
    #[token(".")]
    DOT,
    #[token(",")]
    COMMA,
    #[token("+")]
    PLUS,
    #[token("-")]
    MINUS,
    #[token("*")]
    STAR,
    #[token("/")]
    SLASH,
    #[token("==")]
    EQ,
    #[token("!=")]
    NE,
    #[token("<")]
    LT,
    #[token("<=")]
    LE,
    #[token(">")]
    GT,
    #[token(">=")]
    GE,

    #[regex(r"[a-z][a-zA-Z0-9_]*")]
    ID_LOWER,
    #[regex(r"[0-9]+")]
    NAT_LIT,
    #[regex(r"\s+")]
    WHITESPACE,
    #[regex(r"\(\*[^\*]*\*\)")] // TODO(MH): This is too restrictive.
    COMMENT,

    // Special tokens
    UNKNOWN, // Unknown token, used for error recovery.
    EOF,     // End-of-file.

    BINOP,

    PARAM,
    PARAM_LIST,

    BINOP_EXPR,
    APP_EXPR,
    VAR_EXPR,
    LIT_EXPR,
    PAREN_EXPR,

    LET_MOD,
    LET_VAR,

    LET_EXPR,
    FUN_EXPR,

    ERROR,
    ROOT,
}

impl SyntaxKind {
    pub fn is_trivia(self) -> bool {
        self == Self::WHITESPACE || self == Self::COMMENT
    }

    pub fn as_set(self) -> SyntaxKindSet {
        SyntaxKindSet::singleton(self)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(kind: SyntaxKind) -> Self {
        Self(kind as u16)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum RufusLang {}

impl rowan::Language for RufusLang {
    type Kind = SyntaxKind;
    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        assert!(raw.0 <= SyntaxKind::ROOT as u16);
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }
    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}

#[derive(Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct SyntaxKindSet(u64);

impl SyntaxKindSet {
    pub const fn new() -> Self {
        Self(0)
    }

    pub const fn singleton(kind: SyntaxKind) -> Self {
        Self(1 << (kind as u64))
    }

    pub const fn from<const N: usize>(kinds: [SyntaxKind; N]) -> Self {
        let mut bits = 0;
        let mut i = 0;
        while i < N {
            bits |= 1 << (kinds[i] as u64);
            i += 1;
        }
        Self(bits)
    }

    pub const fn union<const N: usize>(sets: [SyntaxKindSet; N]) -> Self {
        let mut bits = 0;
        let mut i = 0;
        while i < N {
            bits |= sets[i].0;
            i += 1;
        }
        Self(bits)
    }

    pub fn to_vec(self) -> Vec<SyntaxKind> {
        let mut kinds = Vec::new();
        let mut i = 0;
        while i <= SyntaxKind::ROOT as u16 {
            if self.0 & (1 << i) != 0 {
                kinds.push(unsafe { std::mem::transmute::<u16, SyntaxKind>(i) });
            }
            i += 1;
        }
        kinds
    }
}

impl std::fmt::Debug for SyntaxKindSet {
    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        fmt.debug_set().entries(self.to_vec()).finish()
    }
}

pub(crate) trait SyntaxExpecation: Copy + std::fmt::Debug {
    fn contains(&self, kind: SyntaxKind) -> bool;
    fn to_set(&self) -> SyntaxKindSet;
}

impl SyntaxExpecation for SyntaxKind {
    fn contains(&self, kind: SyntaxKind) -> bool {
        *self == kind
    }
    fn to_set(&self) -> SyntaxKindSet {
        SyntaxKindSet::singleton(*self)
    }
}

impl SyntaxExpecation for SyntaxKindSet {
    fn contains(&self, kind: SyntaxKind) -> bool {
        self.0 & (1 << (kind as u64)) != 0
    }

    fn to_set(&self) -> SyntaxKindSet {
        *self
    }
}

pub type SyntaxElement = rowan::SyntaxElement<RufusLang>;
pub type SyntaxNode = rowan::SyntaxNode<RufusLang>;
pub type SyntaxToken = rowan::SyntaxToken<RufusLang>;

pub fn dump_syntax(root: SyntaxNode, include_ws: bool) -> String {
    fn go(node: SyntaxNode, buffer: &mut String, indent: &mut String, include_ws: bool) {
        buffer.push_str(&format!("{}{:?}\n", indent, node));
        indent.push_str("  ");
        for child in node.children_with_tokens() {
            match child {
                SyntaxElement::Node(node) => go(node, buffer, indent, include_ws),
                SyntaxElement::Token(token) => {
                    if include_ws || !token.kind().is_trivia() {
                        buffer.push_str(&format!("{}#{:?}\n", indent, token));
                    }
                }
            }
        }
        indent.truncate(indent.len() - 2);
    }

    let mut buffer = String::new();
    go(root, &mut buffer, &mut String::new(), include_ws);
    // panic!("BUFFER: {}", buffer);
    buffer
}
