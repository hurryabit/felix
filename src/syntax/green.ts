import { Kind } from "./kind";

export class GreenNode {
    readonly kind: Kind;
    readonly length: number; // Length of the text below this node.
    readonly children: readonly GreenChild[];

    constructor(kind: Kind, children: readonly GreenChild[]) {
        this.kind = kind;
        this.length = children.reduce((n, c) => n + c.length, 0);
        this.children = children;
    }

    debug(): string {
        const buffer = [""] as string[];
        this.debugInto(buffer, "");
        buffer.push("");
        return buffer.join("\n");
    }

    debugInto(buffer: string[], indent: string): void {
        buffer.push(`${indent}${this.kind}#${this.length}`);
        indent = indent + "  ";
        this.children.forEach((c) => c.debugInto(buffer, indent));
    }
}

export class GreenToken {
    readonly kind: Kind;
    readonly text: string;

    constructor(kind: Kind, text: string) {
        this.kind = kind;
        this.text = text;
    }

    get length(): number {
        return this.text.length;
    }

    debugInto(buffer: string[], indent: string): void {
        let text = this.text;
        if (/\s/.test(text)) text = encodeURIComponent(text);
        buffer.push(`${indent}${this.kind}#${this.length} :: ${this.text}`);
    }
}

export type GreenChild = GreenNode | GreenToken;

export type Checkpoint = number & { readonly tag: unique symbol };

export class GreenNodeBuilder {
    readonly #interner: Interner;
    readonly #parents: [Kind, number][];
    readonly #children: WithHash<GreenChild>[];

    constructor() {
        this.#interner = new Interner();
        this.#parents = [];
        this.#children = [];
    }

    token(kind: Kind, text: string): void {
        this.#children.push(this.#interner.token(kind, text));
    }

    node(kind: Kind, f: (b: GreenNodeBuilder) => void): void {
        this.#start_node(kind);
        f(this);
        this.#finish_node();
    }

    checkpoint(): Checkpoint {
        return this.#children.length as Checkpoint;
    }

    node_at(checkpoint: Checkpoint, kind: Kind, f: (b: GreenNodeBuilder) => void): void {
        this.#start_node_at(checkpoint, kind);
        f(this);
        this.#finish_node();
    }

    #start_node(kind: Kind) {
        this.#parents.push([kind, this.#children.length]);
    }

    #finish_node() {
        const [kind, start_index] = this.#parents.pop()!;
        const hash_node = this.#interner.node(kind, this.#children, start_index);
        this.#children.push(hash_node);
    }

    #start_node_at(checkpoint: Checkpoint, kind: Kind) {
        if (checkpoint > this.#children.length)
            throw new Error("invalid checkpoint: did it escape its scope?");
        const num_parents = this.#parents.length;
        if (num_parents > 0 && checkpoint < this.#parents[num_parents - 1][1])
            throw new Error("invalid checkpoint: was is captured in a nested scope?");
        this.#parents.push([kind, checkpoint]);
    }

    finish(): GreenNode {
        if (this.#children.length != 1) throw new Error("wrong number of children in finish");
        const { it } = this.#children.pop()!;
        if (!(it instanceof GreenNode)) throw new Error("bad node in finish");
        return it;
    }
}

type WithHash<T> = { readonly hash: number; readonly it: T };

class Interner {
    node(kind: Kind, elements: WithHash<GreenChild>[], start_index: number): WithHash<GreenNode> {
        // TODO(MH): Do the interning.
        console.assert(start_index <= elements.length);
        const children = new Array<GreenChild>(elements.length - start_index);
        let i = children.length;
        while (i != 0) {
            i--;
            children[i] = elements.pop()!.it;
        }
        return {
            hash: 0,
            it: new GreenNode(kind, children),
        };
    }

    token(kind: Kind, text: string): WithHash<GreenToken> {
        // TODO(MH): Do the interning.
        return { hash: 0, it: new GreenToken(kind, text) };
    }
}
