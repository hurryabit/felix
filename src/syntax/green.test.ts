import { describe, expect, test } from "vitest";
import { GreenChild, GreenNode, GreenNodeBuilder, GreenToken } from "./green";
import { Kind } from "./kind";

function token(kind: Kind, text: string): GreenToken {
    return new GreenToken(kind, text);
}

function node(kind: Kind, ...children: GreenChild[]) {
    return new GreenNode(kind, children);
}

describe("GreenNodeBuilder", () => {
    test("simple", () => {
        const b = new GreenNodeBuilder();
        b.node("EXPR_VAR", (b) => {
            b.token("ID_EXPR", "xxx");
        });
        expect(b.finish()).toEqual(node("EXPR_VAR", token("ID_EXPR", "xxx")));
    });

    test("straight", () => {
        const b = new GreenNodeBuilder();
        b.node("EXPR_ABS", (b) => {
            b.token("GR_LAM_LOW", "λ");
            b.node("BINDER", (b) => {
                b.token("ID_EXPR", "xxx");
            });
            b.token("DOT", ".");
            b.node("EXPR_VAR", (b) => {
                b.token("ID_EXPR", "xxx");
            });
        });
        expect(b.finish()).toEqual(
            node(
                "EXPR_ABS",
                token("GR_LAM_LOW", "λ"),
                node("BINDER", token("ID_EXPR", "xxx")),
                token("DOT", "."),
                node("EXPR_VAR", token("ID_EXPR", "xxx")),
            ),
        );
    });

    test("checkpoint", () => {
        const b = new GreenNodeBuilder();
        const cp = b.checkpoint();
        b.node("EXPR_VAR", (b) => b.token("ID_EXPR", "xxx"));
        b.node_at(cp, "EXPR_ADD", (b) => {
            b.token("PLUS", "+");
            b.node("EXPR_VAR", (b) => b.token("ID_EXPR", "yyy"));
        });
        expect(b.finish()).toEqual(
            node(
                "EXPR_ADD",
                node("EXPR_VAR", token("ID_EXPR", "xxx")),
                token("PLUS", "+"),
                node("EXPR_VAR", token("ID_EXPR", "yyy")),
            ),
        );
    });

    test("reused checkpoint", () => {
        const b = new GreenNodeBuilder();
        const cp = b.checkpoint();
        b.node("EXPR_VAR", (b) => b.token("ID_EXPR", "xxx"));
        b.node_at(cp, "EXPR_MUL", (b) => {
            b.token("STAR", "*");
            b.node("EXPR_VAR", (b) => b.token("ID_EXPR", "yyy"));
        });
        b.node_at(cp, "EXPR_ADD", (b) => {
            b.token("PLUS", "+");
            b.node("EXPR_VAR", (b) => b.token("ID_EXPR", "zzz"));
        });
        expect(b.finish()).toEqual(
            node(
                "EXPR_ADD",
                node(
                    "EXPR_MUL",
                    node("EXPR_VAR", token("ID_EXPR", "xxx")),
                    token("STAR", "*"),
                    node("EXPR_VAR", token("ID_EXPR", "yyy")),
                ),
                token("PLUS", "+"),
                node("EXPR_VAR", token("ID_EXPR", "zzz")),
            ),
        );
    });

    test("escaped checkpoint", () => {
        const b = new GreenNodeBuilder();
        let cp;
        b.node("NODE", (b) => {
            b.token("TOKEN", "xxx");
            b.token("TOKEN", "yyy");
            cp = b.checkpoint();
        });
        expect(() => b.node_at(cp!, "NODE", () => {})).toThrowError(/invalid.*escape/);
    });

    test("captured checkpoint", () => {
        const b = new GreenNodeBuilder();
        b.node("NODE", (b) => {
            const cp = b.checkpoint();
            b.token("TOKEN", "xxx");
            b.node("NODE", (b) => {
                expect(() => b.node_at(cp!, "NODE", () => {})).toThrowError(/invalid.*captured/);
            });
        });
    });
});
