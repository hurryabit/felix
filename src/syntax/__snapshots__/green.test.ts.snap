// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GreenNodeBuilder > simple > 
          EXPR_VAR
              ID_EXPR :: text
         1`] = `
"
EXPR_VAR
    ID_EXPR :: text
"
`;
